src/PowerShell/Microsoft.WinGet.DSC/Microsoft.WinGet.DSC.psm1

```diff
@@ -304,4 +310,175 @@ class WinGetSourcesResource
+        $catalogPackage = Get-WinGetPackage -Id $this.Id -Exact
+        if ($null -ne $catalogPackage)
+        {
+            $wingetPackageResource.Installed = $true
+            $wingetPackageResource.Version = $catalogPackage.InstalledVersion.Version
+            $wingetPackageResource.Source = $catalogPackage.DefaultInstallVersion.PackageCatalog.Info.Name
+        }
+
+        return $wingetPackageResource
+    }
+
+    # Test.
+    [bool] Test()
+    {
+        $installedPackage = $this.Get()
+        $ensureInstalled = $this.Ensure -eq [Ensure]::Present
+
+        # This will populate info for Set if needed without another call to Get-WinGetPackage
+        $this.Installed = $installedPackage.Installed
+
+        # Not installed, doesn't have to.
+        if (-not($installedPackage.Installed -or $ensureInstalled))
+        {
+            return $true
+        }
+
+        # Not install, need to ensure installed.
+        # Installed, need to ensure not installed.
+        if ($installedPackage.Installed -ne $ensureInstalled)
+        {
+            return $false
+        }
+
+        # Different sources.
+        if (-not([string]::IsNullOrWhiteSpace($this.Source)))
+        {
+            if ($this.Source -ne $installedPackage.Source)
+            {
+                return $false
+            }
+        }
+
+        $catalogPackage = Get-WinGetPackage -Id $this.Id -Exact
+
+        # If no version is given, see if the latests is installed.
+        if ([string]::IsNullOrWhiteSpace($this.Version) -and
+            $catalogPackage.IsUpdateAvailable)
+        {
+            return $false
+        }
+
+        # If there is an specific version, compare with the current installed version.
+        if (-not ([string]::IsNullOrWhiteSpace($this.Version)))
+        {
+            $compareResult = $catalogPackage.InstalledVersion.CompareToVersion($this.Version)
+            if ($compareResult -ne 'Equal')
+            {
+                return $false
+            }
+        }
+
+        # For now this is all.
+        return $true
+    }
+
+    # Set.
+    [void] Set()
+    {
+        Assert-WinGetCommand "Install-WinGetPackage"
+        Assert-WinGetCommand "Uninstall-WinGetPackage"
+
+        if (-not $this.Test())
+        {
+            $hashArgs = @{
+                Id = $this.Id
+                Exact = $true
+                Mode = 'Silent'
+            }
+            
+            if ($this.Ensure -eq [Ensure]::Present)
+            {
+                if (-not([string]::IsNullOrWhiteSpace($this.Source)))
+                {
+                    $hashArgs.Add("Source", $this.Source)
+                }
+
+                if ($this.Installed)
+                {
+                    $catalogPackage = Get-WinGetPackage -Id $this.Id -Exact
+
+                    if ([string]::IsNullOrWhiteSpace($this.Version))
+                    {
+                        Update-WinGetPackage @hashArgs

```

Comment:
Again, I don't see how not specifying a version means "the latest version" rather than "I don't care what the version is".